src/Admin/Model/UserAdmin.php

```diff
@@ -251,4 +252,19 @@ protected function configureFormFields(FormMapper $formMapper): void
             ->end()
         ;
     }
+
+    private function canImpersonate(): bool
+    {
+        if (!$this->getConfigurationPool()->getContainer()->getParameter('sonata.user.impersonating')) {
+            return false;
+        }
+        $role = 'ROLE_ALLOWED_TO_SWITCH';
+        $firewallMap = $this->getConfigurationPool()->getContainer()->get('security.firewall.map', ContainerInterface::NULL_ON_INVALID_REFERENCE);
+        $firewallConfig = $firewallMap ? $firewallMap->getFirewallConfig($this->getRequest()) : null;
+        if ($firewallConfig && null !== $switchUserConfig = $firewallConfig->getSwitchUser()) {
+            $role = $switchUserConfig['role'];
+        }
+
+        return $this->isGranted($role);
+    }
```

Comment:
We should really refrain from accessing the container like this IMO. Besides, the role is not going to change at runtime. IMO this should be done with configuration, and the configuration node could be changed automatically if not set, by reading the configuration of the security bundle, like so: https://symfony.com/doc/4.1/bundles/prepend_extension.html